src/librustdoc/test.rs

```diff
@@ -410,12 +450,28 @@ impl Collector {
             current_header: None,
             cratename: cratename,
             opts: opts,
+            code_blocks: block_lines,
+            filename: filename,
+        }
+    }
+
+    fn get_line_from_key(&mut self, key: &String) -> u32 {
+        let (line, need_removal) = if let Some(l) = self.code_blocks.get_mut(key) {
+            let need_removal = l.len() > 1;
+            (l.pop().unwrap_or(1), need_removal)
+        } else {
+            return 1;
+        };
+        if need_removal {
+            self.code_blocks.remove(key);
         }
+        line
```

Comment:
I don't quite understand how this hash table works. It looks like it stores a mapping from the complete test case text to the line number it appears on, but I'm not clear on what the vector is for.

I'd guess it's for duplicate test cases, but what mechanism pushes the vectors in the same order they are popped later? Maybe there's an easier way to represent this.